File: src/articulation/pinocchio_model.cpp
#ifdef _USE_PINOCCHIO
#include "pinocchio_model.h"
#include <pinocchio/algorithm/aba.hpp>
#include <pinocchio/algorithm/crba.hpp>
#include <pinocchio/algorithm/joint-configuration.hpp>
#include <pinocchio/algorithm/rnea.hpp>

#define ASSERT(exp, info)                                                                         \
  if (!(exp)) {                                                                                   \
    throw std::runtime_error((info));                                                             \
  }

namespace sapien {
std::unique_ptr<PinocchioModel> PinocchioModel::fromURDFXML(std::string const &urdf) {
  auto m = std::unique_ptr<PinocchioModel>(new PinocchioModel);
  pinocchio::urdf::buildModelFromXML(urdf, m->model);
  m->data = pinocchio::Data(m->model);
  m->indexS2P.setIdentity(m->model.nv);
  return m;
}

Eigen::VectorXd PinocchioModel::posS2P(const Eigen::VectorXd &qext) {
  Eigen::VectorXd qint(model.nq);
  uint32_t count = 0;
  for (size_t N = 0; N < QIDX.size(); ++N) {
    auto start_idx = QIDX[N];
    switch (NQ[N]) {
    case 0:
      break;
    case 1:
      qint[start_idx] = qext[count];
      break;
    case 2:
      qint[start_idx] = std::cos(qext[count]);
      qint[start_idx + 1] = std::sin(qext[count]);
      break;
    default:
      throw std::runtime_error(
          "Unsupported joint in computation. Currently support: fixed, revolute, prismatic");
    }
    count += NV[N];
  }
  ASSERT(count == qext.size(), "posS2P failed");
  return qint;
}

Eigen::VectorXd PinocchioModel::posP2S(const Eigen::VectorXd &qint) {
  Eigen::VectorXd qext(model.nv);

  uint32_t count = 0;
  for (size_t N = 0; N < QIDX.size(); ++N) {
    auto start_idx = QIDX[N];
    switch (NQ[N]) {
    case 0:
      break;
    case 1:
      qext[count] = qint[start_idx];
      break;
    case 2:
      qext[count] = std::atan2(qint[start_idx + 1], qint[start_idx]);
      break;
    default:
      throw std::runtime_error(
          "Unsupported joint in computation. Currently support: fixed, revolute, prismatic");
    }
    count += NV[N];
  }
  ASSERT(count == model.nv, "posP2S failed");
  return qext;
}

void PinocchioModel::setJointOrder(std::vector<std::string> names) {
  Eigen::VectorXi v(model.nv);
  size_t count = 0;
  for (auto &name : names) {
    auto i = model.getJointId(name);
    if (i == model.njoints) {
      throw std::invalid_argument("invalid names in setJointOrder");
    }
    auto size = model.nvs[i];
    auto qi = model.idx_vs[i];
    for (int s = 0; s < size; ++s) {
      v[count++] = qi + s;
    }
  }
  ASSERT(count == model.nv, "setJointOrder failed");
  indexS2P = Eigen::PermutationMatrix<Eigen::Dynamic, Eigen::Dynamic>(v);

  QIDX = Eigen::VectorXi(names.size());
  NQ = Eigen::VectorXi(names.size());
  NV = Eigen::VectorXi(names.size());
  for (size_t N = 0; N < names.size(); ++N) {
    auto i = model.getJointId(names[N]);
    if (i == model.njoints) {
      throw std::invalid_argument("invalid names in setJointOrder");
    }
    NQ[N] = model.nqs[i];
    NV[N] = model.nvs[i];
    QIDX[N] = model.idx_qs[i];
  }
}

void PinocchioModel::setLinkOrder(std::vector<std::string> names) {
  linkIdx2FrameIdx = {};
  Eigen::VectorXi v(names.size());
  for (auto &name : names) {
    auto i = model.getFrameId(name, pinocchio::BODY);
    if (i == model.nframes) {
      throw std::invalid_argument("invalid names in setLinkOrder");
    }
    linkIdx2FrameIdx.push_back(i);
  }
}

Eigen::MatrixXd PinocchioModel::getRandomConfiguration() {
  return posP2S(pinocchio::randomConfiguration(model));
}

void PinocchioModel::computeForwardKinematics(const Eigen::VectorXd &qpos) {
  pinocchio::forwardKinematics(model, data, posS2P(qpos));
}

physx::PxTransform PinocchioModel::getLinkPose(uint32_t index) {
  ASSERT(index < linkIdx2FrameIdx.size(), "link index out of bound");
  auto frame = linkIdx2FrameIdx[index];
  auto parentJoint = model.frames[frame].parent;
  auto link2joint = model.frames[frame].placement;
  auto joint2world = data.oMi[parentJoint];

  auto link2world = joint2world * link2joint;
  auto P = link2world.translation();
  auto Q = Eigen::Quaterniond(link2world.rotation());
  return {physx::PxVec3(P.x(), P.y(), P.z()), physx::PxQuat(Q.x(), Q.y(), Q.z(), Q.w())};
}

void PinocchioModel::computeFullJacobian(const Eigen::VectorXd &qpos) {
  pinocchio::computeJointJacobians(model, data, posS2P(qpos));
}

Eigen::Matrix<double, 6, Eigen::Dynamic> PinocchioModel::getLinkJacobian(uint32_t index,
                                                                         bool local) {
  ASSERT(index < linkIdx2FrameIdx.size(), "link index out of bound");
  auto frameIdx = linkIdx2FrameIdx[index];
  auto jointIdx = model.frames[frameIdx].parent;

  auto link2joint = model.frames[frameIdx].placement;
  auto joint2world = data.oMi[jointIdx];
  auto link2world = joint2world * link2joint;

  Eigen::Matrix<double, 6, Eigen::Dynamic> J(6, model.nv);
  J.fill(0);

  pinocchio::getJointJacobian(model, data, jointIdx, pinocchio::ReferenceFrame::WORLD, J);
  if (local) {
    J = link2world.toActionMatrixInverse() * J;
  }
  // permute Jacobin to SAPIEN format
  return J * indexS2P;
}

Eigen::Matrix<double, 6, Eigen::Dynamic>
PinocchioModel::computeSingleLinkLocalJacobian(Eigen::VectorXd const &qpos, uint32_t index) {
  ASSERT(index < linkIdx2FrameIdx.size(), "link index out of bound");
  auto frameIdx = linkIdx2FrameIdx[index];
  auto jointIdx = model.frames[frameIdx].parent;
  auto link2joint = model.frames[frameIdx].placement;

  Eigen::Matrix<double, 6, Eigen::Dynamic> J(6, model.nv);
  J.fill(0);
  pinocchio::computeJointJacobian(model, data, posS2P(qpos), jointIdx, J);
  return link2joint.toActionMatrixInverse() * J * indexS2P;
}

Eigen::MatrixXd PinocchioModel::computeGeneralizedMassMatrix(const Eigen::VectorXd &qpos) {
  pinocchio::crba(model, data, posS2P(qpos));
  data.M.triangularView<Eigen::StrictlyLower>() =
      data.M.transpose().triangularView<Eigen::StrictlyLower>();
  return indexS2P.transpose() * data.M * indexS2P;
}

Eigen::MatrixXd PinocchioModel::computeCoriolisMatrix(const Eigen::VectorXd &qpos,
                                                      const Eigen::VectorXd &qvel) {
  return indexS2P.transpose() *
         pinocchio::computeCoriolisMatrix(model, data, posS2P(qpos), indexS2P * qvel) * indexS2P;
}

Eigen::VectorXd PinocchioModel::computeInverseDynamics(const Eigen::VectorXd &qpos,
                                                       const Eigen::VectorXd &qvel,
                                                       const Eigen::VectorXd &qacc) {
  return indexS2P.transpose() *
         pinocchio::rnea(model, data, posS2P(qpos), indexS2P * qvel, indexS2P * qacc);
}

Eigen::VectorXd PinocchioModel::computeForwardDynamics(const Eigen::VectorXd &qpos,
                                                       const Eigen::VectorXd &qvel,
                                                       const Eigen::VectorXd &qf) {
  return indexS2P.transpose() *
         pinocchio::aba(model, data, posS2P(qpos), indexS2P * qvel, indexS2P * qf);
}

std::tuple<Eigen::VectorXd, bool, Eigen::Matrix<double, 6, 1>>
PinocchioModel::computeInverseKinematics(uint32_t linkIdx, physx::PxTransform const &pose,
                                         double eps, int maxIter, double dt, double damp) {
  ASSERT(linkIdx < linkIdx2FrameIdx.size(), "link index out of bound");
  auto frameIdx = linkIdx2FrameIdx[linkIdx];
  auto jointIdx = model.frames[frameIdx].parent;
  pinocchio::SE3 l2w;
  l2w.translation({pose.p.x, pose.p.y, pose.p.z});
  l2w.rotation(Eigen::Quaterniond(pose.q.w, pose.q.x, pose.q.y, pose.q.z).toRotationMatrix());
  auto l2j = model.frames[frameIdx].placement;
  pinocchio::SE3 oMdes = l2w * l2j.inverse();

  Eigen::VectorXd q = pinocchio::neutral(model);

  pinocchio::Data::Matrix6x J(6, model.nv);
  J.setZero();
  bool success = false;
  typedef Eigen::Matrix<double, 6, 1> Vector6d;
  Vector6d err;
  Eigen::VectorXd v(model.nv);

  for (int i = 0;; i++) {
    pinocchio::forwardKinematics(model, data, q);
    const pinocchio::SE3 dMi = oMdes.actInv(data.oMi[jointIdx]);
    err = pinocchio::log6(dMi).toVector();
    if (err.norm() < eps) {
      success = true;
      break;
    }
    if (i >= maxIter) {
      success = false;
      break;
    }
    pinocchio::computeJointJacobian(model, data, q, jointIdx, J);
    pinocchio::Data::Matrix6 JJt;
    JJt.noalias() = J * J.transpose();
    JJt.diagonal().array() += damp;
    v.noalias() = -J.transpose() * JJt.ldlt().solve(err);
    q = pinocchio::integrate(model, q, v * dt);
  }
  return {posP2S(q), success, err};
}

} // namespace sapien

#endif

File: python/py_package/wrapper/pinocchio_model.py
import numpy as np
import warnings
from ..pysapien import Pose
from ..pysapien.physx import PhysxArticulation
import platform
from .urdf_exporter import export_kinematic_chain_urdf

try:
    import pinocchio

    class PinocchioModel:
        def __init__(self, urdf_string, gravity):
            self.model: pinocchio.Model = pinocchio.buildModelFromXML(urdf_string)
            self.model.gravity.vector[:] = [*gravity, 0, 0, 0]
            self.data = pinocchio.Data(self.model)

        def set_joint_order(self, names):
            v = np.zeros(self.model.nv, dtype=np.int64)
            count = 0
            for name in names:
                i = self.model.getJointId(name)
                if i >= self.model.njoints:
                    raise ValueError(f"invalid joint name {name}")
                size = self.model.nvs[i]
                qi = self.model.idx_vs[i]
                for s in range(size):
                    v[count] = qi + s
                    count += 1

            self.index_s2p = v

            self.index_p2s = np.zeros_like(self.index_s2p)
            for s, p in enumerate(self.index_s2p):
                self.index_p2s[p] = s

            self.QIDX = np.zeros(
                len(names), dtype=np.int64
            )  # joint position start index
            self.NQ = np.zeros(len(names), dtype=np.int64)  # joint position dim
            self.NV = np.zeros(len(names), dtype=np.int64)  # joint velocity dim
            for N in range(len(names)):
                i = self.model.getJointId(names[N])
                if i >= self.model.njoints:
                    raise ValueError(f"invalid joint name {name}")
                self.NQ[N] = self.model.nqs[i]
                self.NV[N] = self.model.nvs[i]
                self.QIDX[N] = self.model.idx_qs[i]

        def set_link_order(self, names):
            v = []
            for name in names:
                i = self.model.getFrameId(name, pinocchio.BODY)
                if i >= self.model.nframes:
                    raise ValueError(f"invalid joint name {name}")

                v.append(i)

            self.link_id_to_frame_index = np.array(v, dtype=np.int64)

        def q_p2s(self, qint):
            qext = np.zeros(self.model.nv)
            count = 0
            for N in range(len(self.QIDX)):
                start_idx = self.QIDX[N]
                if self.NQ[N] == 1:
                    qext[count] = qint[start_idx]
                elif self.NQ[N] == 2:
                    qext[count] = np.arctan2(qint[start_idx + 1], qint[start_idx])
                elif self.NQ[N] > 2:
                    raise ValueError(
                        f"Unsupported joint in computation. Currently support: fixed, revolute, prismatic"
                    )
                count += self.NV[N]

            assert count == len(self.NV)
            return qext

        def q_s2p(self, qext):
            qint = np.zeros(self.model.nq)
            count = 0
            for N in range(len(self.QIDX)):
                start_idx = self.QIDX[N]
                if self.NQ[N] == 1:
                    qint[start_idx] = qext[count]
                elif self.NQ[N] == 2:
                    qint[start_idx] = np.cos(qext[count])
                    qint[start_idx + 1] = np.sin(qext[count])
                elif self.NQ[N] > 2:
                    raise ValueError(
                        f"Unsupported joint in computation. Currently support: fixed, revolute, prismatic"
                    )
                count += self.NV[N]

            assert count == len(self.NV)
            return qint

        def get_random_qpos(self):
            return self.q_p2s(pinocchio.randomConfiguration(self.model))

        def compute_forward_kinematics(self, qpos):
            """
            Compute and cache forward kinematics. After computation, use get_link_pose to retrieve the computed pose for a specific link.
            """

            pinocchio.forwardKinematics(self.model, self.data, self.q_s2p(qpos))

        def get_link_pose(self, index):
            """
            Given link index, get link pose (in articulation base frame) from forward kinematics. Must be called after compute_forward_kinematics.
            """
            assert 0 <= index < len(self.link_id_to_frame_index)
            frame = int(self.link_id_to_frame_index[index])
            parent_joint = self.model.frames[frame].parent
            link2joint = self.model.frames[frame].placement
            joint2world = self.data.oMi[parent_joint]
            link2world = joint2world * link2joint
            p = link2world.translation
            q = link2world.rotation
            T = np.eye(4)
            T[:3, :3] = q
            T[:3, 3] = p
            return Pose(T)

        def compute_full_jacobian(self, qpos):
            """
            Compute and cache Jacobian for all links
            """
            pinocchio.computeJointJacobians(self.model, self.data, self.q_s2p(qpos))

        def get_link_jacobian(self, index, local=False):
            """
            Given link index, get the Jacobian. Must be called after compute_full_jacobian.

            Args:
              link_index: index of the link
              local: True for world(spatial) frame; False for link(body) frame
            """
            assert 0 <= index < len(self.link_id_to_frame_index)
            frame = int(self.link_id_to_frame_index[index])
            parent_joint = self.model.frames[frame].parent

            link2joint = self.model.frames[frame].placement
            joint2world = self.data.oMi[parent_joint]
            link2world = joint2world * link2joint

            J = pinocchio.getJointJacobian(
                self.model, self.data, parent_joint, pinocchio.ReferenceFrame.WORLD
            )
            if local:
                J = link2world.toActionMatrixInverse() @ J

            return J[:, self.index_s2p]

        def compute_single_link_local_jacobian(self, qpos, index):
            """
            Compute the link(body) Jacobian for a single link. It is faster than compute_full_jacobian followed by get_link_jacobian
            """
            assert 0 <= index < len(self.link_id_to_frame_index)
            frame = int(self.link_id_to_frame_index[index])
            joint = self.model.frames[frame].parent
            link2joint = self.model.frames[frame].placement
            J = pinocchio.computeJointJacobian(
                self.model, self.data, self.q_s2p(qpos), joint
            )
            J = link2joint.toActionMatrixInverse() @ J
            return J[:, self.index_s2p]

        def compute_generalized_mass_matrix(self, qpos):
            return pinocchio.crba(self.model, self.data, self.q_s2p(qpos))[
                self.index_s2p, :
            ][:, self.index_s2p]

        def compute_coriolis_matrix(self, qpos, qvel):
            return pinocchio.computeCoriolisMatrix(
                self.model, self.data, self.q_s2p(qpos), qvel[self.index_p2s]
            )[self.index_s2p, :][:, self.index_s2p]

        def compute_inverse_dynamics(self, qpos, qvel, qacc):
            return pinocchio.rnea(
                self.model,
                self.data,
                self.q_s2p(qpos),
                qvel[self.index_p2s],
                qacc[self.index_p2s],
            )[self.index_s2p]

        def compute_forward_dynamics(self, qpos, qvel, qf):
            return pinocchio.aba(
                self.model,
                self.data,
                self.q_s2p(qpos),
                qvel[self.index_p2s],
                qf[self.index_p2s],
            )[self.index_s2p]

        def compute_inverse_kinematics(
            self,
            link_index,
            pose,
            initial_qpos=None,
            active_qmask=None,
            eps=1e-4,
            max_iterations=1000,
            dt=0.1,
            damp=1e-6,
        ):
            """
            Compute inverse kinematics with CLIK algorithm.
            Details see https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/md_doc_b-examples_i-inverse-kinematics.html
            Also see bug fix from https://github.com/stack-of-tasks/pinocchio/pull/1963/files
            Args:
                link_index: index of the link
                pose: target pose of the link in articulation base frame
                initial_qpos: initial qpos to start CLIK
                active_qmask: dof sized integer array, 1 to indicate active joints and 0 for inactive joints, default to all 1s
                max_iterations: number of iterations steps
                dt: iteration step "speed"
                damp: iteration step "damping"
            Returns:
                result: qpos from IK
                success: whether IK is successful
                error: se3 norm error
            """
            assert 0 <= link_index < len(self.link_id_to_frame_index)
            if initial_qpos is None:
                q = pinocchio.neutral(self.model)
            else:
                q = self.q_s2p(initial_qpos)

            if active_qmask is None:
                mask = np.ones(self.model.nv)
            else:
                mask = np.array(active_qmask)[self.index_p2s]

            mask = np.diag(mask)

            frame = int(self.link_id_to_frame_index[link_index])
            joint = self.model.frames[frame].parent

            T = pose.to_transformation_matrix()
            l2w = pinocchio.SE3()
            l2w.translation[:] = T[:3, 3]
            l2w.rotation[:] = T[:3, :3]

            l2j = self.model.frames[frame].placement
            oMdes = l2w * l2j.inverse()

            best_error = 1e10
            best_q = np.array(q)

            for i in range(max_iterations):
                pinocchio.forwardKinematics(self.model, self.data, q)
                iMd = self.data.oMi[joint].actInv(oMdes)
                err = pinocchio.log6(iMd).vector
                err_norm = np.linalg.norm(err)
                if err_norm < best_error:
                    best_error = err_norm
                    best_q = q

                if err_norm < eps:
                    success = True
                    break

                J = pinocchio.computeJointJacobian(self.model, self.data, q, joint)
                Jlog = pinocchio.Jlog6(iMd.inverse())
                J = -Jlog @ J
                J = J @ mask

                JJt = J @ J.T
                JJt[np.diag_indices_from(JJt)] += damp

                v = -(J.T @ np.linalg.solve(JJt, err))
                q = pinocchio.integrate(self.model, q, v * dt)
            else:
                success = False

            return self.q_p2s(best_q), success, best_error

except ModuleNotFoundError:
    if platform.system() == "Linux":
        from ..pysapien_pinocchio import PinocchioModel
    else:
        warnings.warn(
            "pinnochio package is not installed, robotics functionalities will not be available"
        )
        PinocchioModel = None

except ImportError:
    if platform.system() == "Linux":
        warnings.warn(
            "pinnochio package is broken, fallback to built-in pinocchio. This may be fixed by installing pinocchio via conda instead of pip"
        )
        from ..pysapien_pinocchio import PinocchioModel
    else:
        warnings.warn(
            "pinnochio package is broken, robotics functionalities will not be available"
        )
        PinocchioModel = None


def _create_pinocchio_model(
    articulation: PhysxArticulation, gravity=[0, 0, -9.81]
) -> PinocchioModel:
    xml = export_kinematic_chain_urdf(articulation, force_fix_root=True)
    model = PinocchioModel(xml, gravity)
    model.set_joint_order(
        [f"joint_{j.child_link.index}" for j in articulation.active_joints]
    )
    model.set_link_order([f"link_{l.index}" for l in articulation.links])
    return model


PhysxArticulation.create_pinocchio_model = _create_pinocchio_model
